<?xml version="1.0"?>
<launch>
    
    <!-- ADI 3D ToF Floor Detector using Single ADTF31xx ToF Camera -->


    <!-- Input and Output Modes -->

    <!-- Input Modes, default(0)
        0:ADTF3175D ToF sensor
        2:Rosbag bin file format
        3:Host connected to ADTF31xx
    -->
	<arg name="arg_input_mode" default="3"/>
    <!-- IP of the ADTF31xx board : Applicable only if the input mode is 4 -->
    <arg name="arg_input_sensor_ip"  default="10.43.0.1"/>
    <!-- Output Modes, default(0)
        0 : Disable all outputs
        1 : Enable both AVI and CSV outputs
    -->
	<arg name="arg_output_mode" default="0"/>
    <!-- Input argument for file-io mode (arg_input_mode 2)
        If "arg_input_mode" is 2 then "arg_input_file_name" represents the file name.
        Ex: <arg name="arg_input_file_name" default="$(find adi_3dtof_floor_detector)/../adi_3dtof_input_video_files/adi_3dtof_height_170mm_yaw_0degrees_cam3.bin"/>
    -->
    <arg name="arg_input_file_name" default="$(find adi_3dtof_floor_detector)/../adi_3dtof_input_video_files/adi_3dtof_height_170mm_yaw_0degrees_cam3.bin"/>


    <!-- ADTF31XX Sensor (arg_input_mode = 0) related parameters -->

    <!-- camera_mode
        "lr-native" - 1
        "lr-qnative" - 3
    -->
    <arg name="arg_camera_mode" value="3" />
    <!-- AB threshold value, default(10) -->
    <arg name="arg_ab_threshold" value="10" />
    <!-- Confidence threshold value
	     For Sensor serial number starting with CR and DV: 10,
		 For sensor serial number starting with AM : 25 
	-->
    <arg name="arg_confidence_threshold" value="10" />
    <!-- Configuration fie name of ToF SDK
        "config_adsd3500_adsd3100.json" - Sensor serial number starting with AM
    -->
    <arg name="arg_config_file_name_of_tof_sdk" value="./config/config_adsd3500_adsd3100.json" />


    <!-- Floor Detector related parameters -->

    <!-- Enable RANSAC Floor Detector, default(1)
         1: Activates RANSAC Floor Detector 
         0: Activates Fallback Floor Detector 
    -->
    <arg name="arg_enable_ransac_floor_detection" default="1"/>
    <!-- Distance which determines how close the point must be to the RANSAC plane in order to be selected as inlier, default(0.025mtr) -->
    <arg name="arg_ransac_distance_threshold_mtr" default="0.025"/>
    <!-- Maximum number of RANSAC iterations which is allowed, default(10) -->
    <arg name="arg_ransac_max_iterations" default="10"/>
    <!-- Points with depth value lesser than the given threshold are considered, default(1.5mtr) -->
    <arg name="arg_discard_distance_threshold_mtr_" default="1.5"/>

    <!-- Fallback Floor Detector related parameters -->

    <!-- Set the camera height -->
    <arg name="arg_camera_height_from_floor_in_mtr"  default="0.17"/>
    <!-- Enable Fallback Floor Detector if RANSAC Floor Detector fails, default(1)
         Make sure that "arg_enable_ransac_floor_detection" is set to 1. 
    -->
    <arg name="arg_enable_fallback_floor_detection" default="1"/>
    <!-- Floor height offset for 3D Yw based floor detection, default(0.1mtr) -->
    <arg name="arg_fallback_floor_height_offset_mtr"  default="0.1"/>

    <!-- Output Topics related parameters -->

    <!-- Enable compression for Depth, AB, Floor Mask Output Topics.
    Currenlty, point cloud compression is not supported, default(0) -->
    <arg name="arg_enable_compression_op_image_topics" default="0"/>

    <!-- TF related parametrs -->

    <!--  map
          ^
          |
      camera_device
          ^
          |
      camera_optical
     -->

    <!-- Adjust the prefix based on the actual value -->
    <arg name="ns_prefix_cam" default="cam1" />
    <!--  Camera frame name (forward is +ve X axis) -->
    <arg name="cam_base_frame" default="$(eval arg('ns_prefix_cam') + '_adtf31xx')" />
    <!--  Camera optical frame name (forward is +ve Z axis) -->
    <arg name="cam_base_frame_optical" default="$(eval arg('ns_prefix_cam') + '_adtf31xx_optical')" />
    <!--  Camera's parent frame name -->
    <arg name="cam_parent_frame" value="map"/>
    <!--  Camera's child frame name -->
    <arg name="cam_child_frame" value="$(arg cam_base_frame_optical)"/>

    <!--  Camera position wrt map  -->

    <!--  Camera position X -->
    <arg name="cam_pos_x" value="0.0"/>
    <!--  Camera position Y -->
    <arg name="cam_pos_y" value="0.0"/>
    <!--  Camera position Z -->
    <arg name="cam_pos_z" value="$(arg arg_camera_height_from_floor_in_mtr)"/>
    <!-- Side Tilt: Left side down:-ve  Right side down: +ve -->
    <arg name="cam_roll" value="0"/>
	<!-- Vertical Tilt, down +ve radian, up -ve radian
        Restriction on Pitch, Accepted Range (-1.5708, 1.5708)
        If |Pitch| > 1.57, output from TF transform should be taken as cam_pitch = 3.14 - extracted pitch_angle
    -->
    <arg name="cam_pitch" value="0"/>
    <!-- Horizontal Tilt: Left: +ve, Right: -ve -->
    <arg name="cam_yaw" value="0"/>

    <!-- Fixed conversion - Optical Frame: X-Right,Y-Down,Z-Front to Base Frame : X-Front,Y-Left,Z-Up conversion -->
    <arg name="cam_optical_to_base_roll" value="-1.57"/>
    <arg name="cam_optical_to_base_pitch" value="0"/>
    <arg name="cam_optical_to_base_yaw" value="-1.57"/>

    <!-- Launch Camera Node -->
    <group ns="$(arg ns_prefix_cam)">
      <node name="adi_3dtof_floor_detector_node" pkg="adi_3dtof_floor_detector" type="adi_3dtof_floor_detector_node" output="screen" cwd="node" required="true">
            <param name="param_camera_link" value="$(arg cam_base_frame)"/>
            <param name="param_optical_camera_link" value="$(arg cam_base_frame_optical)"/>
            <param name="param_input_sensor_mode" value="$(arg arg_input_mode)" />
            <param name="param_input_sensor_ip" value="$(arg arg_input_sensor_ip)" />
            <param name="param_output_sensor_mode" value="$(arg arg_output_mode)" />
            <param name="param_input_file_name" value="$(arg arg_input_file_name)"/>

            <param name="param_camera_mode" value="$(arg arg_camera_mode)"/>
            <param name="param_ab_threshold" value="$(arg arg_ab_threshold)"/>
            <param name="param_confidence_threshold" value="$(arg arg_confidence_threshold)" />
            <param name="param_config_file_name_of_tof_sdk" value="$(arg arg_config_file_name_of_tof_sdk)" /> 

            <param name="param_enable_ransac_floor_detection" value="$(arg arg_enable_ransac_floor_detection)" />
            <param name="param_ransac_distance_threshold_mtr" value="$(arg arg_ransac_distance_threshold_mtr)" />
            <param name="param_ransac_max_iterations" value="$(arg arg_ransac_max_iterations)" />
            <param name="param_discard_distance_threshold_mtr_" value="$(arg arg_discard_distance_threshold_mtr_)" />

            <param name="param_enable_fallback_floor_detection" value="$(arg arg_enable_fallback_floor_detection)" />
            <param name="param_fallback_floor_height_offset_mtr" value="$(arg arg_fallback_floor_height_offset_mtr)" />
            
            <param name="param_enable_compression_op_image_topics" value="$(arg arg_enable_compression_op_image_topics)" />

        </node>

        <!-- Transform of Camera Optical and Camera Base frame, position is same as camera base, only change in the rotation -->
        <node pkg="tf2_ros" type="static_transform_publisher" name="$(eval arg('cam_base_frame_optical') + '_tf')" 
                args="0 0 0 $(arg cam_optical_to_base_yaw) $(arg cam_optical_to_base_pitch) $(arg cam_optical_to_base_roll) $(arg cam_base_frame) $(arg cam_child_frame)" />

        <!-- Transform for Camera Base frame and MAP -->
        <node pkg="tf2_ros" type="static_transform_publisher" name="$(eval arg('cam_base_frame') + '_tf')" 
                args="$(arg cam_pos_x) $(arg cam_pos_y) $(arg cam_pos_z) $(arg cam_yaw) $(arg cam_pitch) $(arg cam_roll) $(arg cam_parent_frame) $(arg cam_base_frame)" />

        <!-- Dynamic Reconfigure -->
        <node name="adi_3dtof_floor_detector_dynparam_node" pkg="dynamic_reconfigure" type="dynparam" args="set_from_parameters adi_3dtof_floor_detector_node">
        <param name = "enable_ransac_floor_detection" type="bool" value="$(arg arg_enable_ransac_floor_detection)" />
        <param name = "ransac_max_iterations" type="int" value="$(arg arg_ransac_max_iterations)" />
        <param name = "ransac_distance_threshold_mtr" type="double" value="$(arg arg_ransac_distance_threshold_mtr)" />

        <param name = "enable_fallback_floor_detection" type="bool" value="$(arg arg_enable_fallback_floor_detection)" />
        <param name = "fallback_floor_height_offset_mtr" type="double" value="$(arg arg_fallback_floor_height_offset_mtr)" />

        <param name = "ab_threshold" type="int" value="$(arg arg_ab_threshold)" />
        <param name = "confidence_threshold" type="double" value="$(arg arg_confidence_threshold)" />

        </node>

    </group>


</launch>
